### PR TITLE
Add test coverage

### PR DESCRIPTION
**Pull request by kumdeepakgit**
Add test coverage for writer pkg

```
        github.com/dineshba/tf-summarize/reader         coverage: 0.0% of statements
=== RUN   TestVersionArg
=== RUN   TestVersionArg/when_tf-summarize_is_passed_'-v'
--- PASS: TestVersionArg (0.00s)
    --- PASS: TestVersionArg/when_tf-summarize_is_passed_'-v' (0.00s)
=== RUN   TestTFSummarize
=== RUN   TestTFSummarize/when_tf-summarize_is_passed_'"./tf-summarize-test_-md_example/tfplan.json"'
=== RUN   TestTFSummarize/when_tf-summarize_is_passed_'"cat_example/tfplan.json_|_./tf-summarize-test_-md"'
=== RUN   TestTFSummarize/when_tf-summarize_is_passed_'"cat_example/tfplan.json_|_./tf-summarize-test_-html"'
=== RUN   TestTFSummarize/when_tf-summarize_is_passed_'"cat_example/tfplan.json_|_./tf-summarize-test_-md_-html"'
--- PASS: TestTFSummarize (0.02s)
    --- PASS: TestTFSummarize/when_tf-summarize_is_passed_'"./tf-summarize-test_-md_example/tfplan.json"' (0.00s)
    --- PASS: TestTFSummarize/when_tf-summarize_is_passed_'"cat_example/tfplan.json_|_./tf-summarize-test_-md"' (0.00s)
    --- PASS: TestTFSummarize/when_tf-summarize_is_passed_'"cat_example/tfplan.json_|_./tf-summarize-test_-html"' (0.00s)
    --- PASS: TestTFSummarize/when_tf-summarize_is_passed_'"cat_example/tfplan.json_|_./tf-summarize-test_-md_-html"' (0.00s)
PASS
coverage: 0.0% of statements
ok      github.com/dineshba/tf-summarize        0.470s  coverage: 0.0% of statements
=== RUN   TestJSONParser_Parse_Success
--- PASS: TestJSONParser_Parse_Success (0.00s)
=== RUN   TestJSONParser_Parse_InvalidJSON
--- PASS: TestJSONParser_Parse_InvalidJSON (0.00s)
=== RUN   TestNewJSONParser
--- PASS: TestNewJSONParser (0.00s)
=== RUN   TestCreateParser_JSONFile
--- PASS: TestCreateParser_JSONFile (0.00s)
=== RUN   TestCreateParser_Stdin
--- PASS: TestCreateParser_Stdin (0.00s)
=== RUN   TestCreateParser_BinaryFile
--- PASS: TestCreateParser_BinaryFile (0.00s)
=== RUN   TestCreateParser_InvalidFileName
--- PASS: TestCreateParser_InvalidFileName (0.00s)
PASS
coverage: 45.5% of statements
ok      github.com/dineshba/tf-summarize/parser 0.008s  coverage: 45.5% of statements
=== RUN   TestResourceChangeColor
--- PASS: TestResourceChangeColor (0.00s)
=== RUN   TestGetAllResourceChanges
--- PASS: TestGetAllResourceChanges (0.00s)
=== RUN   TestGetAllOutputChanges
--- PASS: TestGetAllOutputChanges (0.00s)
=== RUN   TestResourceChangeSuffix
--- PASS: TestResourceChangeSuffix (0.00s)
=== RUN   TestResourceChangeColorAndSuffixImport
--- PASS: TestResourceChangeColorAndSuffixImport (0.00s)
=== RUN   TestFilterNoOpResources
--- PASS: TestFilterNoOpResources (0.00s)
PASS
coverage: 93.7% of statements
ok      github.com/dineshba/tf-summarize/terraformstate 0.008s  coverage: 93.7% of statements
=== RUN   TestCreateTreeForEmptyResourceChanges
--- PASS: TestCreateTreeForEmptyResourceChanges (0.00s)
=== RUN   TestCreateTreeForOneResourceChanges
--- PASS: TestCreateTreeForOneResourceChanges (0.00s)
=== RUN   TestCreateTreeWithQuotesInResources
--- PASS: TestCreateTreeWithQuotesInResources (0.00s)
=== RUN   TestCreateTreeForOneResourceChangesMultiLevel
--- PASS: TestCreateTreeForOneResourceChangesMultiLevel (0.00s)
=== RUN   TestCreateTreeForTwoResourceChangesNoOverlap
--- PASS: TestCreateTreeForTwoResourceChangesNoOverlap (0.00s)
=== RUN   TestCreateTreeForTwoResourceChangesOverlap
--- PASS: TestCreateTreeForTwoResourceChangesOverlap (0.00s)
PASS
coverage: 62.1% of statements
ok      github.com/dineshba/tf-summarize/tree   0.009s  coverage: 62.1% of statements
=== RUN   TestHTMLWriterWithMockFileSystem
--- PASS: TestHTMLWriterWithMockFileSystem (0.00s)
=== RUN   TestJSONWriter
--- PASS: TestJSONWriter (0.00s)
=== RUN   TestSeparateTree_Write
=== RUN   TestSeparateTree_Write/Drawable_True
=== RUN   TestSeparateTree_Write/Drawable_False
=== RUN   TestSeparateTree_Write/Error_Handling
=== RUN   TestSeparateTree_Write/Error_Handling/Write_Error
=== RUN   TestSeparateTree_Write/Error_Handling/Tree_Writer_Error
=== RUN   TestSeparateTree_Write/Error_Handling/Final_Newline_Write_Error
--- PASS: TestSeparateTree_Write (0.00s)
    --- PASS: TestSeparateTree_Write/Drawable_True (0.00s)
    --- PASS: TestSeparateTree_Write/Drawable_False (0.00s)
    --- PASS: TestSeparateTree_Write/Error_Handling (0.00s)
        --- PASS: TestSeparateTree_Write/Error_Handling/Write_Error (0.00s)
        --- PASS: TestSeparateTree_Write/Error_Handling/Tree_Writer_Error (0.00s)
        --- PASS: TestSeparateTree_Write/Error_Handling/Final_Newline_Write_Error (0.00s)
=== RUN   TestTableWriter_Write_NoMarkdown
--- PASS: TestTableWriter_Write_NoMarkdown (0.00s)
=== RUN   TestTableWriter_Write_WithMarkdown
--- PASS: TestTableWriter_Write_WithMarkdown (0.00s)
=== RUN   TestTableWriter_NoChanges
--- PASS: TestTableWriter_NoChanges (0.00s)
=== RUN   TestTreeWriter_Write_DrawableTrue
--- PASS: TestTreeWriter_Write_DrawableTrue (0.00s)
=== RUN   TestTreeWriter_Write_NonDrawable
--- PASS: TestTreeWriter_Write_NonDrawable (0.00s)
=== RUN   TestTreeWriter_Write_NonDrawable_PrintTreeError
--- PASS: TestTreeWriter_Write_NonDrawable_PrintTreeError (0.00s)
=== RUN   TestTreeWriter_Write_PrintTreeError
--- PASS: TestTreeWriter_Write_PrintTreeError (0.00s)
=== RUN   TestTreeWriter_Write_EmptyChanges
--- PASS: TestTreeWriter_Write_EmptyChanges (0.00s)
=== RUN   TestHasOutputChanges
=== RUN   TestHasOutputChanges/No_changes
=== RUN   TestHasOutputChanges/Has_changes
=== RUN   TestHasOutputChanges/All_empty
--- PASS: TestHasOutputChanges (0.00s)
    --- PASS: TestHasOutputChanges/No_changes (0.00s)
    --- PASS: TestHasOutputChanges/Has_changes (0.00s)
    --- PASS: TestHasOutputChanges/All_empty (0.00s)
=== RUN   TestRemoveANSI
=== RUN   TestRemoveANSI/String_without_ANSI_escape
=== RUN   TestRemoveANSI/String_with_ANSI_escape
=== RUN   TestRemoveANSI/String_with_multiple_ANSI_escapes
=== RUN   TestRemoveANSI/Empty_string
--- PASS: TestRemoveANSI (0.00s)
    --- PASS: TestRemoveANSI/String_without_ANSI_escape (0.00s)
    --- PASS: TestRemoveANSI/String_with_ANSI_escape (0.00s)
    --- PASS: TestRemoveANSI/String_with_multiple_ANSI_escapes (0.00s)
    --- PASS: TestRemoveANSI/Empty_string (0.00s)
PASS
coverage: 87.4% of statements
ok      github.com/dineshba/tf-summarize/writer 0.011s  coverage: 87.4% of statements
```

### CODE REVIEW

File: writer/separate_tree_test.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+|	|---example2(-)
+
+
+`
+
+	assert.Equal(t, expectedOutput, removeANSI(buf.String()))
+}
+
+// Mock writer that returns an error after a certain number of writes
+type controlledErrorWriter struct {
```
Comment (dineshba) [inline writer/separate_tree_test.go:111]: Is it possible to use gomock for this?

File: writer/util.go [hunk 2]
```diff
@@ -1,12 +1,20 @@
...
+	"io/fs"
+	"regexp"
+)
 
 // Embed the templates directory in the compiled binary.
 //
 //go:embed templates
 var templates embed.FS
 
+var getFS = func() fs.FS {
```
Comment (dineshba) [inline writer/util.go:14]: Any reason for declaring getFS as variable?

File: writer/json_test.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+								"name": "old_sg",
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	for i, changes := range mockResourceChangesArray {
```
Comment (dineshba) [inline writer/json_test.go:105]: It is better to always group all inputs into one struct for tests. Is https://www.digitalocean.com/community/tutorials/how-to-write-unit-tests-in-go-using-go-test-and-the-testing-package#step-4-mdash-writing-table-driven-tests-in-go useful here?

**Review by dineshba — COMMENTED** [review]
Thanks for the PR @kumdeepakgit ... Have added few minor comments

**Comment by dineshba** [comment]
Also, looks there is a test failure: https://github.com/dineshba/tf-summarize/actions/runs/11233918381/job/31284281419?pr=78#step:8:83

File: writer/separate_tree_test.go (hunk 1 above)
Comment (kumdeepakgit) [reply writer/separate_tree_test.go]: Yes, it's a good idea, Have updated to use gomock to mock the io writer interface and handle the expected calls of the writer func

image

File: writer/util.go (hunk 2 above)
Comment (kumdeepakgit) [reply writer/util.go:14]: I'm unable to replace the `templates` var directly in test cases due to this error
`cannot use fstest.MapFS{…} (value of type fstest.MapFS) as embed.FS value in assignment`

So, This is the one option that I could find, to convert the embed.FS to FS interface

File: writer/json_test.go (hunk 3 above)
Comment (kumdeepakgit) [reply writer/json_test.go]: Done

**Comment by kumdeepakgit** [comment]
> Also, looks there is a test failure: https://github.com/dineshba/tf-summarize/actions/runs/11233918381/job/31284281419?pr=78#step:8:83

@dineshba  Yes, there were some flaky test cases, have updated those

**Review by dineshba — APPROVED** [review]

**Pull request merged**